src/backend/distributed/utils/background_jobs.c

```diff
@@ -310,6 +330,214 @@ CitusBackgroundTaskQueueMonitorErrorCallback(void *arg)
+		(void) WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT |
+						 WL_EXIT_ON_PM_DEATH,
+						 delay_ms, WAIT_EVENT_PG_SLEEP);
+		ResetLatch(MyLatch);
+
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CreateNewExecutor returns the handle to new task executor.
+ */
+static BackgroundWorkerHandle *
+CreateNewExecutor(BackgroundTask *task, dsm_segment **seg, char *databaseName,
+				  char *userName)
+{
+	/* try to allocate a new task executor */
+	BackgroundWorkerHandle *handle = StartCitusBackgroundTaskExecutor(databaseName,
+																	  userName,
+																	  task->command,
+																	  task->taskid, seg);
+	return handle;
+}
+
+
+/*
+ * AssignRunnableTaskToNewExecutor tries to assign given runnable task to a new task executor. It reports the assignment status as return value.
+ */
+static bool
+AssignRunnableTaskToNewExecutor(BackgroundTask *runnableTask,
+								HTAB *backgroundExecutorHandles,
+								QueueMonitorExecutionContext *queueMonitorExecutionContext)
+{
+	Assert(runnableTask && runnableTask->status == BACKGROUND_TASK_STATUS_RUNNABLE);
+
+	if (NewExecutorExceedsCitusLimit(queueMonitorExecutionContext))
+	{
+		/* escape if we hit citus executor limit */
+		return false;
+	}
+
+	char *databaseName = get_database_name(MyDatabaseId);
+	char *userName = GetUserNameFromId(runnableTask->owner, false);
+
+	/* try to create new executor and make it alive during queue monitor lifetime */
+	MemoryContext oldContext = MemoryContextSwitchTo(queueMonitorExecutionContext->ctx);
+	dsm_segment *seg = NULL;
+	BackgroundWorkerHandle *handle = CreateNewExecutor(runnableTask, &seg, databaseName,
+													   userName);
+	MemoryContextSwitchTo(oldContext);
+
+	if (NewExecutorExceedsPgMaxWorkers(handle, queueMonitorExecutionContext))
+	{
+		/* escape if we hit pg worker limit */
+		return false;
+	}
+
+	/* assign the allocated executor to the runnable task and increment total executor count */
+	bool handleEntryFound = false;
+	BackgroundExecutorHashEntry *handleEntry = hash_search(backgroundExecutorHandles,
+														   &runnableTask->taskid,
+														   HASH_ENTER, &handleEntryFound);
+	Assert(!handleEntryFound);
+	handleEntry->handle = handle;
+	handleEntry->seg = seg;
+
+	/* make message alive during queue monitor lifetime */
+	oldContext = MemoryContextSwitchTo(queueMonitorExecutionContext->ctx);
+	handleEntry->message = makeStringInfo();
+	MemoryContextSwitchTo(oldContext);
+
+	/* set runnable task's status as running */
+	runnableTask->status = BACKGROUND_TASK_STATUS_RUNNING;
+	UpdateBackgroundTask(runnableTask);
+	UpdateBackgroundJob(runnableTask->jobid);
+
+	queueMonitorExecutionContext->currentExecutorCount++;
+
+	ereport(LOG, (errmsg("task jobid/taskid started: %ld/%ld",
+						 runnableTask->jobid, runnableTask->taskid)));
+
+	return true;
+}
+
+
+/*
+ * AssignRunnableTasks tries to assign all runnable tasks to a new task executor. If an assignment fails, it stops in case we hit some limitation.
```

Comment:
nit: optimal line length can be debated, but these comment lines are objectively quite long. Can you hard-wrap them to something reasonable (80/100/120 characters)?

(same for function above)